cpp/src/lance/arrow/type.cc

```diff
@@ -43,36 +80,88 @@ ::arrow::Result<std::string> ToLogicalType(std::shared_ptr<::arrow::DataType> dt
+    {"uint64", ::arrow::uint64()},
+    {"halffloat", ::arrow::float16()},
+    {"float", ::arrow::float32()},
+    {"double", ::arrow::float64()},
+    {"string", ::arrow::utf8()},
+    {"binary", ::arrow::binary()},
+    {"large_string", ::arrow::large_utf8()},
+    {"large_binary", ::arrow::large_binary()},
+    {"date32:day", ::arrow::date32()},
+    {"date64:ms", ::arrow::date64()},
+};
+
+::arrow::Result<::arrow::TimeUnit::type> TimeUnitFromLogicalType(
+    const ::arrow::util::string_view& unit) {
+  using Unit = ::arrow::TimeUnit;
+  if (unit == "s") {
+    return Unit::SECOND;
+  } else if (unit == "ms") {
+    return Unit::MILLI;
+  } else if (unit == "us") {
+    return Unit::MICRO;
+  } else if (unit == "ns") {
+    return Unit::NANO;
+  };
+  return ::arrow::Status::Invalid(fmt::format("Unsupported TimeUnit: {}", unit.to_string()));
+}
+
+::arrow::Result<std::shared_ptr<::arrow::DataType>> TimeFromLogicalType(
+    const ::arrow::util::string_view& logical_type) {
+  auto components = ::arrow::internal::SplitString(logical_type, ':');
+  if (components.size() != 2) {
+    return ::arrow::Status::Invalid(
+        fmt::format("Invalid timestamp string: {}", logical_type.to_string()));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto unit, TimeUnitFromLogicalType(components[1]));
+  if (components[0] == "timestamp") {
+    return ::arrow::timestamp(unit);
+  } else if (components[0] == "time32") {
+    return ::arrow::time32(unit);
+  } else if (components[0] == "time64") {
+    return ::arrow::time64(unit);
+  }
+  return ::arrow::Status::Invalid(
+      fmt::format("Invalid temporal logical type: {}", logical_type.to_string()));
+};
+
 ::arrow::Result<std::shared_ptr<::arrow::DataType>> FromLogicalType(
     ::arrow::util::string_view logical_type) {
-  // TODO: optimize this lookup table?
-  if (logical_type == "bool") {
-    return ::arrow::boolean();
-  } else if (logical_type == "int8") {
-    return ::arrow::int8();
-  } else if (logical_type == "uint8") {
-    return ::arrow::utf8();
-  } else if (logical_type == "int16") {
-    return ::arrow::int16();
-  } else if (logical_type == "uint16") {
-    return ::arrow::uint16();
-  } else if (logical_type == "int32") {
-    return ::arrow::int32();
-  } else if (logical_type == "uint32") {
-    return ::arrow::uint32();
-  } else if (logical_type == "int64") {
-    return ::arrow::int64();
-  } else if (logical_type == "uint64") {
-    return ::arrow::uint64();
-  } else if (logical_type == "float") {
-    return ::arrow::float32();
-  } else if (logical_type == "double") {
-    return ::arrow::float64();
-  } else if (logical_type == "string") {
-    return ::arrow::utf8();
-  } else if (logical_type == "binary") {
-    return ::arrow::binary();
-  } else if (logical_type.starts_with("dict")) {
+  const auto& it = kPrimitiveTypeMap.find(logical_type.to_string());
+  if (it != kPrimitiveTypeMap.end()) {
+    return it->second;
+  }
+
+  if (logical_type.starts_with("time")) {
```

Comment:
isn't there an integer type id we can use instead of needing to do a string comparison?

Reply:
The logical type is stored within lance `protobuf Field`.